packages/file/lib/src/io_overrides.dart

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io' as io
+    show
+        ConnectionTask,
+        IOOverrides,
+        ServerSocket,
+        Socket,
+        stderr,
+        Stdin,
+        stdin,
+        Stdout,
+        stdout;
+
+import 'interface.dart';
+import 'io.dart' as io;
+
+/// Internal base class for [IOOverrides] implementations backed by a
+/// specified [FileSystem].
+///
+/// Operations that are not filesystem-related fall back to the normal
+/// `dart:io` implementations.
+///
+/// Do not use this with a [LocalFileSystem]; doing so would cause overridden
+/// operations to create infinite recursion loops.  To reduce the likelihood of
+/// such problems, this class is not exported and is not intended to be used
+/// directly.
+abstract class FileSystemIOOverrides implements io.IOOverrides {
```

Comment:
Rather than implementing `io.IOOverrides`, have a `FileSystemOverrides` that only includes the properties relevent to a file system (`stdio` is not a file system concept).

Then have the `dart:io`-supporting `FileSystemOverrides` create an underlying `IOOverrides` that it passes to `dart:io`, and the memory-file-system's `FileSystemOverrides` has its own implementation.

In general, I'd have preferred an API that is separate from that of `dart:io`, so it can be implemented on any platform, and therefore not refer to `dart:io` in the public API here, if avoidable. 
That ship has sailed since `File` implements `io.File` rather than wrapping it. (Oh well, maybe  another package can do that.)